src/linux/mod.rs

```diff
@@ -41,14 +42,96 @@ pub struct Segment<'a> {
 }
 
 impl<'a> Segment<'a> {
+    fn phdr(&self) -> &'a Phdr {
+        unsafe { self.phdr.as_ref().unwrap() }
+    }
+
+    /// You must pass this segment's `SharedLibrary` or else this is wild UB.
+    unsafe fn data(&self, shlib: &SharedLibrary<'a>) -> Option<&'a [u8]> {
+        let phdr = self.phdr();
+        (shlib.addr as usize)
+            .checked_add(phdr.p_vaddr as usize)
+            .map(|avma| slice::from_raw_parts(avma as *const u8, phdr.p_memsz as usize))
+    }
+
     fn is_load(&self) -> bool {
-        unsafe {
-            let hdr = self.phdr.as_ref().unwrap();
-            match hdr.p_type {
-                libc::PT_LOAD => true,
-                _ => false,
+        self.phdr().p_type == libc::PT_LOAD
+    }
+
+    fn is_note(&self) -> bool {
+        self.phdr().p_type == libc::PT_NOTE
+    }
+
+    /// Parse the contents of a `PT_NOTE` segment.
+    ///
+    /// Returns a triple of
+    ///
+    /// 1. The `NT_*` note type.
+    /// 2. The note name.
+    /// 3. The note descriptor payload.
+    ///
+    /// You must pass this segment's `SharedLibrary` or else this is wild UB.
+    unsafe fn notes(
+        &self,
+        shlib: &SharedLibrary<'a>,
+    ) -> impl Iterator<Item = (libc::Elf32_Word, &'a [u8], &'a [u8])> {
+        assert!(self.is_note());
+
+        let phdr = self.phdr();
+
+        let alignment = std::cmp::min(phdr.p_align as usize, 4);
+        let align = move |data: &'a [u8]| {
+            // Same logic as in `gimli` which took it from `readelf`.
+            if alignment != 4 && alignment != 8 {
+                return None;
             }
-        }
+
+            let ptr = data.as_ptr() as usize;
+            let diff = ptr % alignment;
+            if diff <= data.len() {
+                Some(&data[diff..])
+            } else {
+                None
+            }
+        };
+
+        let mut data = self.data(shlib).unwrap_or(&[]);
+
+        iter::from_fn(move || {
+            // Each entry in a `PT_NOTE` segment begins with a
+            // fixed-size header `Nhdr`.
+            //
+            // We always use an Nhdr32 here as 64-bit notes have not
+            // been observed in practice.
+            let nhdr_size = mem::size_of::<Nhdr32>();
+            if data.len() < nhdr_size {
+                data = &[];
+                return None;
+            }
+            let nhdr = (data.as_ptr() as *const Nhdr32).as_ref().unwrap();
+            data = &data[nhdr_size..];
+
+            // It is followed by a name of size `n_namesz`.
+            let name_size = nhdr.n_namesz as usize;
+            if data.len() < name_size {
+                data = &[];
+                return None;
+            }
+            let name = &data[..name_size];
```

Comment:
These 5 lines are similar to `let name = data.get(..name_size)?;`, with the difference being that it doesn't set `data = &[]` on error, but then the `align` below doesn't set `data = &[]` on error either so does that difference matter?